cmd/kubeadm/app/discovery/file/file.go

```diff
@@ -53,58 +51,44 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
 		return nil, err
 	}
 
-	// This is the cluster object we've got from the cluster-info kubeconfig file
-	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
+	var kubeconfig *clientcmdapi.Config
 
-	// Create a new kubeconfig object from the given, just copy over the server and the CA cert
-	// We do this in order to not pick up other possible misconfigurations in the clusterinfo file
-	kubeconfig := kubeconfigutil.CreateBasic(
-		defaultCluster.Server,
-		clustername,
-		"", // no user provided
-		defaultCluster.CertificateAuthorityData,
-	)
-	// load pre-existing client certificates
-	if config.Contexts[config.CurrentContext] != nil && len(config.AuthInfos) > 0 {
-		user := config.Contexts[config.CurrentContext].AuthInfo
-		authInfo, ok := config.AuthInfos[user]
-		if !ok || authInfo == nil {
-			return nil, errors.Errorf("empty settings for user %q", user)
-		}
-		if len(authInfo.ClientCertificateData) == 0 && len(authInfo.ClientCertificate) != 0 {
-			clientCert, err := ioutil.ReadFile(authInfo.ClientCertificate)
-			if err != nil {
-				return nil, err
-			}
-			authInfo.ClientCertificateData = clientCert
-		}
-		if len(authInfo.ClientKeyData) == 0 && len(authInfo.ClientKey) != 0 {
-			clientKey, err := ioutil.ReadFile(authInfo.ClientKey)
-			if err != nil {
-				return nil, err
-			}
-			authInfo.ClientKeyData = clientKey
-		}
+	// If the discovery file config contains a authentication credentials
+	if kubeconfigutil.HasAuthenticationCredentials(config) {
+		klog.V(1).Info("[discovery] Using authentication credentials from the discovery file for validating TLS connection")
 
-		if len(authInfo.ClientCertificateData) == 0 || len(authInfo.ClientKeyData) == 0 {
-			return nil, errors.New("couldn't read authentication info from the given kubeconfig file")
+		// Use the discovery file config for starting the join process
+		kubeconfig = config
+
+		// We should ensure that all the authentication info are embedded in config file, so everything will work also when
+		// the kubeconfig file will be stored in /etc/kubernetes/boostrap-kubelet.conf
+		if err := kubeconfigutil.EnsureAuthenticationInfoAreEmbedded(kubeconfig); err != nil {
```

Comment:
can the logic here reach a case where `HasAuthenticationCredentials()` returns `true` for a token in the kubeconfig, but then we want to load the certs on disk into the file?

Reply:
kubeconfig is validated before this call, and this ensures only one authentication method is configured. Nevertheless, if in the future this `HasAuthenticationCredentials` it will be called somewhere else, I don't see problems because this function and `EnsureAuthenticationInfoAreEmbedded` have clearly separated scope

Reply:
i guess in the case that i'm talking about the function `EnsureAuthenticationInfoAreEmbedded()` will fail with an `ioutil.ReadFile(authInfo.ClientCertificate)` error.

if the user has provided a kubeconfig with a CA and token it will pass the `HasAuthenticationCredentials` check, but then fail the `EnsureAuthenticationInfoAreEmbedded` check because the kubeconfig does not have client cert/key. so it seems to me this case is unsupported.

i think if `EnsureAuthenticationInfoAreEmbedded` fails here we should validate the token in the kubeconfig and only then exit the branch with an error.

Reply:
Clarified. EnsureAuthenticationInfoAreEmbedded does not fail in case of Kubeconfig file with token